core/templates/services/audio-player.service.ts

```diff
@@ -16,162 +16,235 @@
-    var _currentTrackFilename = null;
-    var _currentTrack = null;
-
-    var _autoplayAudioEventEmitter = new EventEmitter();
-
-    var _load = function(
-        filename, successCallback, errorCallback) {
-      if (filename !== _currentTrackFilename) {
-        AssetsBackendApiService.loadAudio(
-          ContextService.getExplorationId(), filename)
-          .then(function(loadedAudiofile) {
-            var blobUrl = URL.createObjectURL(loadedAudiofile.data);
-            _currentTrack = ngAudio.load(blobUrl);
-            _currentTrackFilename = filename;
-
-            // Directive ngAudio doesn't seem to provide any way of detecting
-            // when native audio object has finished loading -- see
-            // https://github.com/danielstern/ngAudio/issues/139. It seems
-            // that after creating an ngAudio object, the native audio
-            // object is asynchronously loaded. So we use a timeout
-            // to grab native audio.
-            // TODO(tjiang11): Look for a better way to handle this.
-            $timeout(function() {
-              // _currentTrack could be null if the learner stops audio
-              // shortly after loading a new card or language. In such
-              // cases, we do not want to attempt setting the 'onended'
-              // property of the audio.
-              if (_currentTrack !== null &&
-                _currentTrack.audio !== undefined) {
-                _currentTrack.audio.onended = function() {
-                  _currentTrack = null;
-                  _currentTrackFilename = null;
-                  AudioTranslationManagerService
-                    .clearSecondaryAudioTranslations();
-                };
-              }
-            }, 100);
-
-            successCallback();
-          }, function(reason) {
-            errorCallback(reason);
-          });
-      }
-    };
+import { EventEmitter, Injectable, NgZone } from '@angular/core';
+import { AudioFile } from 'domain/utilities/audio-file.model';
+import { AudioTranslationManagerService } from 'pages/exploration-player-page/services/audio-translation-manager.service';
+import { AssetsBackendApiService } from './assets-backend-api.service';
+import { ContextService } from './context.service';
+import { Howl } from 'howler';
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { interval, Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
 
-    var _play = function() {
-      if (_currentTrack) {
-        _currentTrack.play();
-      }
-    };
+@Injectable({
+  providedIn: 'root'
+})
+export class AudioPlayerService {
+  private _currentTrackFilename: string | null = null;
+  private _currentTrack: Howl | null = null;
+  private _lastPausePos: number | null = null;
+  private loadingTrack = false;
+  private _updateViewEventEmitter = new EventEmitter();
+  private _autoplayAudioEventEmitter = new EventEmitter();
+  private _stopIntervalSubject = new Subject();
```

Comment:
type?

Reply:
All of them are void emitters... they don't emit any value... they just emit. Typescript auto implies the type here. For example, if you try to change the value of loadingTrack here to 4 or `Hello`, TS will raise an error. So I think I will as types for the subjects and eventemitters.

Reply:
Ok